In [1]:
%run setup.ipynb

In [2]:
df = pd.read_csv(f'{RESULTS_PATH}/user_support_BASE.csv')

In [4]:
print(df.shape)
df.head()

(28531, 9)


from                                    to  \
0  758ec809-2969-47ef-b5db-3f9520cc7726  55b732b1-48be-4af0-b318-f7415adb04b2   
1  a6e1196b-613d-4cfe-9d08-8cc5b8a0dd6b  e1e8d77c-b4cd-42e2-8957-0640d4f7a200   
2  758ec809-2969-47ef-b5db-3f9520cc7726  e57948bb-7a75-4c2a-a956-a9578eb8f3c0   
3  fae942d2-d644-4870-9792-221d17e6cd6a  baf79279-a045-4e94-8e9e-8d684694d67a   
4  a5564a0b-f5b0-4c91-b813-adca235331fb  69c527ce-9ce8-4e83-bb6a-1199d753f544   

                                             subject  \
0  Re: [user-support] URGENTE problema per access...   
1          [user-support] Fwd: Richiesta Accesso HPC   
2        Re: [user-support] GERDA @ cnaf after flood   
3               Re: [user-support] Fwd: Account CNAF   
4  Re: [user-support] [Junodatacenter] JUNO StoRM...   

                                             content        type  \
0  Ciao Marco,\r\n\r\nvogliamo provare a fare un ...  text/plain   
1  Ciao Daniele,\r\n\r\nCi confermi la creazione ...  text/plain   
2  Ciao tutti,\r\n\r\nvi informo che la coda di g...  text/plain   
3  Ciao Luca,\r\n\r\nperfetto, grazie della confe...  text/plain   
4  Dear Daniele,\n\nThank you for letting us know...  text/plain   

                              date  \
0  Thu, 12 Dec 2019 16:21:51 +0100   
1  Thu, 12 Jan 2023 11:21:40 +0100   
2  Thu, 12 Apr 2018 10:13:33 +0200   
3  Mon, 26 Oct 2020 11:50:53 +0100   
4  Fri, 25 Sep 2020 22:25:39 +0800   

                                                  cc  \
0           ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0']   
1           ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0']   
2  ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0', '7f0f...   
3           ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0']   
4                                               ['']   

                                                  id  \
0  <26c572e8-8278-36d8-d337-0683fecaa918@cnaf.inf...   
1  <86C8CDAF-A7B0-437E-B9DC-417D7CA5D304@cnaf.inf...   
2  <f96583f7-31ef-21d0-0cc0-9a3a26af1708@cnaf.inf...   
3  <2cf128fd-be86-2f8a-2a31-8bee827a62a3@cnaf.inf...   
4  <44f18486-a63a-47ee-f0a2-b2a9550c556b@ihep.ac.cn>   

                                              parent  
0   <0A442A0B-2706-4A33-84D8-58681DD50DE8@gmail.com>  
1                                                NaN  
2  <alpine.LRH.2.11.1804111337560.12376@gerda-log...  
3  <6cb270dd-2286-fa42-6aca-f35a8903d1a0@ge.infn.it>  
4  <c03a8a8b-3b6d-466f-2a1b-561272c91f98@cnaf.inf...

In [5]:
# we are checking the missing values in the content or in the subject of the emails 
non_string_subject = df[~df['subject'].apply(lambda x: isinstance(x, str))]['subject']
non_string_content = df[~df['content'].apply(lambda x: isinstance(x, str))]['content']

print("Non-string values in 'subject':", non_string_subject)
print("Non-string values in 'content':", non_string_content)

Non-string values in 'subject': Series([], Name: subject, dtype: object)
Non-string values in 'content': 490      NaN
510      NaN
894      NaN
1942     NaN
2575     NaN
4040     NaN
4300     NaN
4996     NaN
5838     NaN
7525     NaN
8506     NaN
9431     NaN
11354    NaN
11585    NaN
12078    NaN
12954    NaN
15677    NaN
15973    NaN
15980    NaN
18326    NaN
20937    NaN
23685    NaN
26926    NaN
27018    NaN
27234    NaN
Name: content, dtype: object


## We check how many emails have the label tag

In [6]:
count_cnaf_farming = df[df['subject'].str.contains(r'\[CNAF farming\]')].shape[0]
print("Numero di email con [CNAF farming] nel subject:", count_cnaf_farming)

Numero di email con [CNAF farming] nel subject: 1454


In [7]:
count_cnaf_farming = df[df['subject'].str.contains(r'\[CNAF net\]')].shape[0]
print("Numero di email con [CNAF net] nel subject:", count_cnaf_farming)

Numero di email con [CNAF net] nel subject: 245


In [8]:
count_cnaf_farming = df[df['subject'].str.contains(r'\[CNAF storage\]')].shape[0]
print("Numero di email con [CNAF storage] nel subject:", count_cnaf_farming)

Numero di email con [CNAF storage] nel subject: 776


In [9]:
count_user_support = df[df['subject'].str.contains(r'\[user-support\]')].shape[0]
print("Numero di email con [user-support] nel subject:", count_user_support)

Numero di email con [user-support] nel subject: 27494


In [10]:
df['content'] = df['content'].fillna('')
# vediamo quante email contengono la stringa "[sysop]" nel campo "subject"
count_sysop_subject = df[df['subject'].str.contains(r'sysop', case=False)].shape[0]
count_sysop_content = df[df['content'].str.contains(r'sysop', case=False)].shape[0]
 
print("Numero di email con 'sysop' nel subject:", count_sysop_subject)
print("Numero di email con 'sysop' nel content:", count_sysop_content)
 
# totale presenze sysop
total_count_sysop = count_sysop_subject + count_sysop_content
print("Numero totale di email con 'sysop' nel subject o nel content:", total_count_sysop)

Numero di email con 'sysop' nel subject: 7
Numero di email con 'sysop' nel content: 1391
Numero totale di email con 'sysop' nel subject o nel content: 1398


In [11]:
count_both_sysop = df[df['subject'].str.contains(r'sysop', case=False) & df['content'].str.contains(r'sysop', case=False)].shape[0]
print("Numero di email con 'sysop' nel subject e 'sysop' nel content contemporaneamente:", count_both_sysop)

Numero di email con 'sysop' nel subject e 'sysop' nel content contemporaneamente: 3


## Define the function to categorize emails 

In [12]:
def categorize_email(email):
    # managing NaN in the "content"
    if not pd.notna(email["content"]):
        # if content is NaN, call the categorize_subject function, so use just the subject to categorize
        return categorize_subject(email["subject"])
    
    if re.search(r'sysop', email["content"], re.IGNORECASE) or re.search(r'sysop', email["subject"], re.IGNORECASE):
        return "sysop"

    # Check for mailing list tags in the subject or content
    tags_subject = re.findall(r'\[([^]]+)\]', email["subject"])
    tags_content = re.findall(r'\[([^]]+)\]', email["content"])
    
    all_tags = tags_subject + tags_content
    
    for tag in all_tags:
        if tag in ["CNAF storage", "CNAF farming", "CNAF net"]:
            return tag
    
    return "without tag"

def categorize_subject(subject):
    if "CNAF storage" in subject:
        return "CNAF storage"
    if "CNAF farming" in subject:
        return "CNAF farming"
    if "CNAF net" in subject:
        return "CNAF net"
    return "without tag"

In [13]:
df['label'] = df.apply(categorize_email, axis = 1)
df['label'].head

<bound method NDFrame.head of 0        without tag
1        without tag
2        without tag
3        without tag
4        without tag
            ...     
28526    without tag
28527    without tag
28528    without tag
28529    without tag
28530    without tag
Name: label, Length: 28531, dtype: object>

In [14]:
class_counts = df['label'].value_counts()
print(class_counts)

label
without tag     24822
CNAF farming     1425
sysop            1395
CNAF storage      729
CNAF net          160
Name: count, dtype: int64


These are the categorized emails. The dataset is unbalanced: for the CNAF net we have just 160 emails.
For the three categories which have 'CNAF' in the label tag, there are less emails than previuosly because the function assigns first of all 'sysop' labels so if an email has double category, it will have 'sysop' tag.

In [15]:
df[:3]

from                                    to  \
0  758ec809-2969-47ef-b5db-3f9520cc7726  55b732b1-48be-4af0-b318-f7415adb04b2   
1  a6e1196b-613d-4cfe-9d08-8cc5b8a0dd6b  e1e8d77c-b4cd-42e2-8957-0640d4f7a200   
2  758ec809-2969-47ef-b5db-3f9520cc7726  e57948bb-7a75-4c2a-a956-a9578eb8f3c0   

                                             subject  \
0  Re: [user-support] URGENTE problema per access...   
1          [user-support] Fwd: Richiesta Accesso HPC   
2        Re: [user-support] GERDA @ cnaf after flood   

                                             content        type  \
0  Ciao Marco,\r\n\r\nvogliamo provare a fare un ...  text/plain   
1  Ciao Daniele,\r\n\r\nCi confermi la creazione ...  text/plain   
2  Ciao tutti,\r\n\r\nvi informo che la coda di g...  text/plain   

                              date  \
0  Thu, 12 Dec 2019 16:21:51 +0100   
1  Thu, 12 Jan 2023 11:21:40 +0100   
2  Thu, 12 Apr 2018 10:13:33 +0200   

                                                  cc  \
0           ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0']   
1           ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0']   
2  ['0327ac73-54cb-4bf6-a9b8-dce5a8b0dbe0', '7f0f...   

                                                  id  \
0  <26c572e8-8278-36d8-d337-0683fecaa918@cnaf.inf...   
1  <86C8CDAF-A7B0-437E-B9DC-417D7CA5D304@cnaf.inf...   
2  <f96583f7-31ef-21d0-0cc0-9a3a26af1708@cnaf.inf...   

                                              parent        label  
0   <0A442A0B-2706-4A33-84D8-58681DD50DE8@gmail.com>  without tag  
1                                                NaN  without tag  
2  <alpine.LRH.2.11.1804111337560.12376@gerda-log...  without tag

In [16]:
df.to_csv(f'{RESULTS_PATH}/user_support2.csv', index=False)